Packing a single executable fat-Jar does not work

Hi team,

We have a runnable fat-jar and when I tried to configure the buildpack to simply pack it and run it, I keep running into the buildpack detecting it as an exploded Jar. In other words, I could not configure it to reach https://github.com/paketo-buildpacks/executable-jar/blob/50bd25d87a7ce37cb10fef4b9084d9489bab071c/executable/executable_jar.go#L56. I forked the repository and commented out the exploded Jar detection `if` and everything worked fine.

## Expected Behavior
Allow packing a single executable fat-Jar.

## Current Behavior
Only performs to pack an exploded Jar.

## Possible Solution
I can see a few possibilities to fix it:
* Adjust the condition that detects whether it is an exploded Jar
* Make the buildpack participation conditions disjuctive, that is, if a runnable jar is detected, then it is a jar, if not, require a MANIFEST.MF and treat it as exploded Jar

## Steps to Reproduce
1. Create an executable far-Jar
2. Define project.toml as:
```
[_]
schema-version = "0.2"
id = "com.test.jar"
name = "Test Fat Jar"

[[io.buildpacks.group]]
uri = "gcr.io/paketo-buildpacks/sap-machine"

[[io.buildpacks.group]]
uri = "gcr.io/paketo-buildpacks/syft"

[[io.buildpacks.group]]
uri = "gcr.io/paketo-buildpacks/executable-jar"

[[io.buildpacks.build.env]]
name = "BP_JVM_VERSION"
value = "21"
```
3. [this step is required only due to the activation conditions] Extract jar's MANIFEST.MF into META-INF/MANIFEST.MF
4. Build the package by running:
```
pack build example:latest \
  --descriptor project.toml \
  --builder paketobuildpacks/builder:base
```
6. Run the image by:
```
docker run --name example --rm example
```
7. The following error is received:
```
Error: Could not find or load main class [class name in the extracted MANIFEST.MF]
Caused by: java.lang.ClassNotFoundException: [class name in the extracted MANIFEST.MF]
```

Looking at the generated image layer, the contents of the layer seem as expected:
```
Permission     UID:GID       Size  Filetree                                     
drwxrwxrwx   1000:1000      76 MB  └── workspace                                
drwxr-xr-x   1000:1000       68 B      ├── META-INF                             
-rw-rw----   1000:1000       68 B      │   └── MANIFEST.MF                      
-rw-r--r--   1000:1000      76 MB      └── example.jar
```

## Motivations
<!--- How has this issue affected you? What are you trying to accomplish? What is the impact? Providing context helps us come up with a solution that is most useful in the real world. -->
I cannot use the buildpack to package my jar file and for me renders the buildpack unusable, however, as I have read in the tests, this is one of the desired usecases of using this buildpack. It would be great if we can either fix the issue or clarify the detailed conditions in which it can be used in such a scenario.

--- a/README.md
+++ b/README.md
@@ -4,7 +4,7 @@ The Paketo Buildpack for Executable JAR is a Cloud Native Buildpack that contrib
 
 ## Behavior
 
-This buildpack will participate if all the following conditions are met:
+This buildpack will participate if any the following conditions are met:
 
 * `<APPLICATION_ROOT>/META-INF/MANIFEST.MF` contains a `Main-Class` entry
 * `<APPLICATION_ROOT>/**/*.jar` exists and that JAR has a `/META-INF/MANIFEST.MF` file which contains a `Main-Class` entry

--- a/executable/detect.go
+++ b/executable/detect.go
@@ -18,6 +18,7 @@ package executable
 
 import (
 	"fmt"
+	"reflect"
 
 	"github.com/buildpacks/libcnb"
 	"github.com/paketo-buildpacks/libjvm"
@@ -65,15 +66,20 @@ func (d Detect) Detect(context libcnb.DetectContext) (libcnb.DetectResult, error
 		return libcnb.DetectResult{}, fmt.Errorf("unable to read manifest in %s\n%w", context.Application.Path, err)
 	}
 
-	jarGlob, _ := cr.Resolve("BP_EXECUTABLE_JAR_LOCATION")
-	_, props, _ := findExecutableJAR(context.Application.Path, jarGlob)
-
 	if _, ok := m.Get("Main-Class"); ok {
 		d.Logger.Info("PASSED: 'Main-Class' manifest attribute found")
 		result.Plans[0].Provides = append(result.Plans[0].Provides, libcnb.BuildPlanProvide{Name: PlanEntryJVMApplicationPackage})
-	} else if _, ok := props.Get("Main-Class"); ok {
-		d.Logger.Info("PASSED: Jar file with 'Main-Class' manifest attribute found")
-		result.Plans[0].Provides = append(result.Plans[0].Provides, libcnb.BuildPlanProvide{Name: PlanEntryJVMApplicationPackage})
+	} else {
+		jarGlob, _ := cr.Resolve("BP_EXECUTABLE_JAR_LOCATION")
+		execJar, err := LoadExecutableJAR(context.Application.Path, jarGlob)
+		if err != nil {
+			return libcnb.DetectResult{}, fmt.Errorf("unable to load executable JAR\n%w", err)
+		}
+
+		if !reflect.DeepEqual(execJar, ExecutableJAR{}) {
+			d.Logger.Info("PASSED: 'Main-Class' manifest attribute found")
+			result.Plans[0].Provides = append(result.Plans[0].Provides, libcnb.BuildPlanProvide{Name: PlanEntryJVMApplicationPackage})
+		}
 	}
 
 	if cr.ResolveBool("BP_LIVE_RELOAD_ENABLED") {

--- a/executable/detect.go
+++ b/executable/detect.go
@@ -21,7 +21,6 @@ import (
 	"reflect"
 
 	"github.com/buildpacks/libcnb"
-	"github.com/paketo-buildpacks/libjvm"
 	"github.com/paketo-buildpacks/libpak"
 	"github.com/paketo-buildpacks/libpak/bard"
 )
@@ -61,25 +60,15 @@ func (d Detect) Detect(context libcnb.DetectContext) (libcnb.DetectResult, error
 		},
 	}
 
-	m, err := libjvm.NewManifest(context.Application.Path)
+	jarGlob, _ := cr.Resolve("BP_EXECUTABLE_JAR_LOCATION")
+	execJar, err := LoadExecutableJAR(context.Application.Path, jarGlob)
 	if err != nil {
-		return libcnb.DetectResult{}, fmt.Errorf("unable to read manifest in %s\n%w", context.Application.Path, err)
+		return libcnb.DetectResult{}, fmt.Errorf("unable to load executable JAR\n%w", err)
 	}
 
-	if _, ok := m.Get("Main-Class"); ok {
+	if !reflect.DeepEqual(execJar, ExecutableJAR{}) {
 		d.Logger.Info("PASSED: 'Main-Class' manifest attribute found")
 		result.Plans[0].Provides = append(result.Plans[0].Provides, libcnb.BuildPlanProvide{Name: PlanEntryJVMApplicationPackage})
-	} else {
-		jarGlob, _ := cr.Resolve("BP_EXECUTABLE_JAR_LOCATION")
-		execJar, err := LoadExecutableJAR(context.Application.Path, jarGlob)
-		if err != nil {
-			return libcnb.DetectResult{}, fmt.Errorf("unable to load executable JAR\n%w", err)
-		}
-
-		if !reflect.DeepEqual(execJar, ExecutableJAR{}) {
-			d.Logger.Info("PASSED: 'Main-Class' manifest attribute found")
-			result.Plans[0].Provides = append(result.Plans[0].Provides, libcnb.BuildPlanProvide{Name: PlanEntryJVMApplicationPackage})
-		}
 	}
 
 	if cr.ResolveBool("BP_LIVE_RELOAD_ENABLED") {

--- a/executable/detect.go
+++ b/executable/detect.go
@@ -33,7 +33,7 @@ const (
 	PlanEntrySyft                  = "syft"
 )
 
-type Detect struct{
+type Detect struct {
 	Logger bard.Logger
 }
 
@@ -65,8 +65,14 @@ func (d Detect) Detect(context libcnb.DetectContext) (libcnb.DetectResult, error
 		return libcnb.DetectResult{}, fmt.Errorf("unable to read manifest in %s\n%w", context.Application.Path, err)
 	}
 
+	jarGlob, _ := cr.Resolve("BP_EXECUTABLE_JAR_LOCATION")
+	_, props, _ := findExecutableJAR(context.Application.Path, jarGlob)
+
 	if _, ok := m.Get("Main-Class"); ok {
 		d.Logger.Info("PASSED: 'Main-Class' manifest attribute found")
+		result.Plans[0].Provides = append(result.Plans[0].Provides, libcnb.BuildPlanProvide{Name: PlanEntryJVMApplicationPackage})
+	} else if _, ok := props.Get("Main-Class"); ok {
+		d.Logger.Info("PASSED: Jar file with 'Main-Class' manifest attribute found")
 		result.Plans[0].Provides = append(result.Plans[0].Provides, libcnb.BuildPlanProvide{Name: PlanEntryJVMApplicationPackage})
 	}
 

--- a/executable/detect_test.go
+++ b/executable/detect_test.go
@@ -132,6 +132,32 @@ func testDetect(t *testing.T, context spec.G, it spec.S) {
 		})
 	})
 
+	context("META-INF/MANIFEST.MF not found but there is JAR with Main-Class", func() {
+		it.Before(func() {
+			Expect(CreateJAR(filepath.Join(ctx.Application.Path, "a.jar"), map[string]string{"Main-Class": "test.Main"})).To(Succeed())
+		})
+
+		it("requires and provides jvm-application-package", func() {
+			Expect(detect.Detect(ctx)).To(Equal(libcnb.DetectResult{
+				Pass: true,
+				Plans: []libcnb.BuildPlan{
+					{
+						Provides: []libcnb.BuildPlanProvide{
+							{Name: "jvm-application"},
+							{Name: "jvm-application-package"},
+						},
+						Requires: []libcnb.BuildPlanRequire{
+							{Name: "syft"},
+							{Name: "jre", Metadata: map[string]interface{}{"launch": true}},
+							{Name: "jvm-application-package"},
+							{Name: "jvm-application"},
+						},
+					},
+				},
+			}))
+		})
+	})
+
 	context("$BP_LIVE_RELOAD_ENABLED is set", func() {
 		it.Before(func() {
 			Expect(os.Setenv("BP_LIVE_RELOAD_ENABLED", "true")).To(Succeed())

